File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""Rectangle class that represents a geometric rectangle shape."""

class Rectangle:
    """A geometric rectangle shape is represented by this class."""
    def __init__(self, width=0, height=0):
        """A new Rectangle instance is initialized.
    
        Args:
            width (int): The width of the rectangle.
            height (int): The height of the rectangle.
        """
        self.__width = width
        self.__height = height
    
    @property
    def width(self):
        """The width of the rectangle is gotten or set."""
        return self.__width
    
    @width.setter
    def width(self, value):
        """The width of the rectangle is set.
    
        Args:
            value (int): The width value to be set.
    
        Raises:
            TypeError: If the provided width is not an integer.
            ValueError: If the provided width is negative.
        """
        if not type(value) == int:
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")
        self.__width = value
    
    @property
    def height(self):
        """The height of the rectangle is gotten or set."""
        return self.__height
    
    @height.setter
    def height(self, value):
        """The height of the rectangle is set.
    
        Args:
            value (int): The height value to be set.
    
        Raises:
            TypeError: If the provided height is not an integer.
            ValueError: If the provided height is negative.
        """
        if not type(value) == int:
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")
        self.__height = value
    
    def area(self):
        """The area of the rectangle is calculated and returned.
    
        Returns:
            int: The area is calculated.
        """
        area = self.__width * self.__height
        return area
    
    def perimeter(self):
        """The perimeter of the rectangle is calculated and returned.
    
        Returns:
            int: The perimeter is calculated.
        """
        while self.__width != 0 and self.__height != 0:
            perimeter = (self.__width * 2) + (self.__height * 2)
            return perimeter
        return 0
    
    def __str__(self):
        """A string representation of the rectangle is returned.
    
        Returns:
            str: The string representation is returned.
        """
        if self.__width == 0 or self.__height == 0:
            return ""
    
        therectangle = []
        i = 0
        while i < self.__height:
            therectangle.append("#" * self.__width)
            if i != self.__height - 1:
                therectangle.append("\n")
            i += 1
    
        return "".join(therectangle)
    
    def __repr__(self):
        """A string representation of the rectangle object is returned.
    
        Returns:
            str: The string representation is returned.
        """
        therectangle = f"Rectangle({self.__width}, {self.__height})"
        return therectangle

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node
class Rectangle {
  constructor(w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  get width() {
    return this._width;
  }

  set width(w) {
    if (w > 0) {
      this._width = w;
    }
  }

  get height() {
    return this._height;
  }

  set height(h) {
    if (h > 0) {
      this._height = h;
    }
  }

  print() {
    for (let i = 0; i < this._height; i++) {
      let s = '';
      for (let j = 0; j < this._width; j++) {
        s += 'X';
      }
      console.log(s);
    }
  }

  rotate() {
    [this._width, this._height] = [this._height, this._width];
  }

  double() {
    this._width *= 2;
    this._height *= 2;
  }
}

module.exports = Rectangle;

